docs/csharp/programming-guide/concepts/expression-trees/how-to-use-expression-trees-to-build-dynamic-queries.md

```diff
@@ -1,107 +1,266 @@
+- Varying the expression tree passed into the LINQ methods
+- Constructing expression trees using the factory methods at <xref:System.Linq.Expressions.Expression>
+
+## Referencing runtime state from the expression tree
+
+Assuming the LINQ provider supports it, the simplest way to query dynamically is to reference the runtime state directly in the query, via a closed-over variable:
+
+```csharp
+var length = 1;
+var qry = companyNamesSource
+    .Select(x => x.Substring(0, length))
+    .Distinct();
+    
+Console.WriteLine(string.Join(",", qry));
+// prints: C, A, S, W, G, H, M, N, B, T, L, F
+
+length = 2;
+Console.WriteLine(string.Join(",", qry));
+// prints: Co, Al, So, Ci, Wi, Gr, Ad, Hu, Wo, Ma, No, Bl, Tr, Th, Lu, Fo 
+```
+
+Note that the expression tree hasn't been modified; the query returns different values only because the value of `length` has been changed.
+
+## Calling additional LINQ methods
+
+The LINQ methods on <xref:System.Linq.Queryable> which don't execute the query but rather return a new translated query object, generally consist of two steps:
+
+* wrap the current expression tree in a <xref:System.Linq.Expressions.MethodCallExpression> representing the method call, and
+* pass the newly wrapped expression tree back into the provider for optional additional processing.
+
+For example, the source code for <xref:System.Linq.Queryable.Take> looks something like this:
+
+```csharp
+public static IQueryable<TSource> Take<TSource>(this IQueryable<TSource> source, int count) {
+
+    // wrap the original expression in a MethodCallExpression using the Call factory method    
+    var wrapped =
+        Expression.Call(
+            null,
+            CachedReflectionInfo.Take_TSource_2(typeof(TSource)),
+
+            // the original expression
+            source.Expression, 
+
+            Expression.Constant(count)
+        )
+
+    // pass the wrapped expression back to the provider
+    var newQuery = source.Provider.CreateQuery<TSource>(wrapped);
+
+    return newQuery;
+}
+```
+
+Thus, based on runtime state, you can conditionally replace the original query object with the result of these method calls, to get a new query.
+
+```csharp
+bool sortByLength = /* ... */;
+
+var qry = companyNamesSource;
+if (sortByLength) {
+    qry = qry.OrderBy(x => x.Length);
+}
+```
+
+## Varying the expression tree passed into the LINQ methods
+
+You can also choose to pass in different expressions to the LINQ methods, depending on runtime state:
+
+```csharp
+string startsWith = /* ... */;
+string endsWith = /* ... */;
+
+Expression<Func<string, bool>>? expr = null;
+if (!string.IsNullOrEmpty(startsWith) && !string.IsNullOrEmpty(endsWith)) {
+    expr = x => x.StartsWith(startsWith) || x.EndsWith(endsWith);
+} else if (!string.IsNullOrEmpty(startsWith)) {
+    expr = x => x.StartsWith(startsWith);
+} else if (!string.IsNullOrEmpty(endsWith)) {
+    expr = x => x.EndsWith(endsWith);
+} else {
+    expr = x => true;
+}
```

Comment:
I haven't refactored this, but I think this could be made more clear with a pattern matching `switch` expression over the tuple `(startsWith, endsWith)`.

Reply:
Fixed in cc6550ea6ce39020e8838d4ff08fb2fb59232a73